src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/ScenarioCustomMarshaler.cs

```diff
@@ -69,5 +69,164 @@ public void Test()
+        private static RecordValue Marshal(Dictionary<string, object> values, TypeMarshallerCache cache, RecordType type)
+        {
+            var fieldMap = new Dictionary<string, Func<object, FormulaValue>>();
+            foreach (var kv in values)
+            {
+                var fieldName = kv.Key;
+
+                var marshaller = cache.GetMarshaller(kv.Value.GetType());
+                var expectedType = type.GetFieldType(fieldName);
+
+                Assert.True(marshaller.Type.Equals(expectedType));
+
+                fieldMap[fieldName] = (source) => 
+                {
+                    var dict = (Dictionary<string, object>)source;
+                    var fieldValue = dict[fieldName];
+                    
+                    var result = marshaller.Marshal(fieldValue);
+                    return result;
+                };
+            }
+
+            var om = new ObjectMarshaller(type, fieldMap);
+            
+            var value = (RecordValue)om.Marshal(values);
+            return value;
+        }
+
+        [Fact]
+        public void TestDictWithKnownType()
+        {
+            // We can have a RecordValue backed by a dynamic dictionary, with lazy field evaluation.
+            // We just need to know the record type. 
+            var d = new Dictionary<string, object>
+            {
+                { "int", 123 },
+                { "str", "string" },
+                { "test", new TestObj { Field1 = 0, Field2 = 22 } },
+                { "bool", true }
+            };
+
+            var recordTypeTestObj = new RecordType()
+                .Add("Field1", FormulaType.Number)
+                .Add("Field2", FormulaType.Number);
+
+            var recordType = new RecordType()
+                .Add("int", FormulaType.Number)
+                .Add("str", FormulaType.String)
+                .Add("test", recordTypeTestObj)
+                .Add("bool", FormulaType.Boolean);
+
+            var cache = new TypeMarshallerCache();
+            var x = Marshal(d, cache, recordType);
+
+            var engine = new RecalcEngine();
+            engine.UpdateVariable("x", x);
+
+            // runtime casting error when we access this field. 
+            // This also proves access is lazy. 
+            d["bool"] = "error: not a bool";
+
+            d["int"] = 456; // lazy, takes latest values. 
+            var result1 = engine.Eval("x.int");
+            Assert.Equal(456.0, ((NumberValue)result1).Value);
+
+            ((TestObj)d["test"]).Field1 = 12;
+            var result2 = engine.Eval("x.test.Field1");
+            Assert.Equal(12.0, ((NumberValue)result2).Value);
+
+            // It's strongly typed, so we can give design-time errors for missing fields. 
+            var result3 = engine.Check("x.missing");
+            Assert.False(result3.IsSuccess);
+
+            // runtime casting error.             
+            Assert.Throws<AggregateException>(() => engine.Eval("x.bool"));
```

Comment:
Curious, why AggregateException here? Should we be providing a more meaningful exception for runtime marshalling failures, or is that on the implementer of the marshaller?

Reply:
It's a .net thing unrelated to this change - Task generally causes AggregateException to be thrown; `await` will unwrap that for you to the inner exception.  But that means things that call `.Result` may still get an aggregate whereas `await` gets the inner.